### PR TITLE
3178/replace function for event handler

### PR DESCRIPTION
**Pull request by jona-samuel**
This is the replacement function for the webhook event handler. The issue is #3178

### CODE REVIEW

File: privacyidea/lib/utils/__init__.py [hunk 1]
```diff
@@ -1480,3 +1481,44 @@ def parse_string_to_dict(s, split_char=":"):
...
+            "logged_in_user": login,
+            "realm": realm,
+            "surname": surname,
+            "token_owner": givenname,
+            "user_realm": userrealm,
+            "token_serial": token_serial
+        }
+        new_text = text.format_map(defaultdict(str, attributes))
+        return new_text
+    except(ValueError) as err:
```
[the remark this answers is not in the document]
Comment (jona-samuel) [reply privacyidea/lib/utils/__init__.py:1519]: 1d32a9d

File: privacyidea/lib/utils/__init__.py [hunk 2]
```diff
@@ -1480,3 +1481,41 @@ def parse_string_to_dict(s, split_char=":"):
...
+    if tokenowner is not None:
+        surname = tokenowner.info.get("surname")
+        givenname = tokenowner.info.get("givenname")
+        userrealm = tokenowner.realm
+    else:
+        surname = ""
+        givenname = ""
+        userrealm = ""
+
+    if token_serial is not None:
```
Comment (plettich) [inline privacyidea/lib/utils/__init__.py:1503]: You can shorten this to 
`token_serial = token_serial if token_serial else ""`

File: privacyidea/lib/utils/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (plettich) [reply privacyidea/lib/utils/__init__.py:1519]: If a placeholder is not present in the `attributes` dictionary, this throws a `KeyError`.
So the best thing would be to rewrite this to something like:
```
try:
    ...
except (KeyError, ValueError) as e:
    log.warning("Unable to replace placeholder: ({0!s})! Please check the webhooks data option.".format(e))
```

File: tests/test_lib_events.py [hunk 3]
```diff
@@ -2677,3 +2682,84 @@ def test_05_wrong_url(self):
...
+                           "request": req,
+                           "handler_def": {
+                               "options": {"URL":
+                                               'http://test.com',
+                                           "content_type":
+                                               CONTENT_TYPE.JSON,
+                                           "replace":
+                                               True,
+                                           "data":
+                                               '{{token_serial} {token_owner} {user_realm}}'
```
Comment (plettich) [inline tests/test_lib_events.py:2754]: I think the most common error in this case is a misspelled placeholder like `{token_seril}`

**Review by plettich — CHANGES_REQUESTED** [review]

File: tests/test_lib_events.py (hunk 3 above)
Comment (jona-samuel) [reply tests/test_lib_events.py:2754]: 604e4e6

File: privacyidea/lib/utils/__init__.py (hunk 2 above)
Comment (jona-samuel) [reply privacyidea/lib/utils/__init__.py:1503]: 604e4e6

File: tests/test_lib_utils.py [hunk 4]
```diff
@@ -2,7 +2,10 @@
 """
 This tests the package lib.utils
 """
-from .base import MyTestCase
+from flask.testing import EnvironBuilder
+
+from privacyidea.lib.token import init_token
+from .base import MyTestCase, FakeFlaskG
```
[the remark this answers is not in the document]
Comment (jona-samuel) [reply tests/test_lib_utils.py]: 604e4e6

File: doc/eventhandler/webhookhandler.rst [hunk 5]
```diff
@@ -21,4 +21,7 @@ what ever your other application understands as data.
 
 .. note:: You can use placeholder for more flexibility of the webhook. For example: If the user John is logged in
     and the webhook handler text is "This webhook is triggered by {logged_in_user}", than this text is send:
-    "This webhook is triggered by John"
+    "This webhook is triggered by John".
+    Possible please holder are `logged_in_user`, `realm`, `token_owner`, `user_realm` and `token_serial`
+
+.. note:: placeholders are only available with python 3. python 2 will not support this
```
[the remark this answers is not in the document]
Comment (jona-samuel) [reply doc/eventhandler/webhookhandler.rst]: 604e4e6

File: doc/eventhandler/webhookhandler.rst [hunk 6]
```diff
@@ -21,4 +21,7 @@ what ever your other application understands as data.
 
 .. note:: You can use placeholder for more flexibility of the webhook. For example: If the user John is logged in
     and the webhook handler text is "This webhook is triggered by {logged_in_user}", than this text is send:
-    "This webhook is triggered by John"
```
[the remark this answers is not in the document]
Comment (jona-samuel) [reply doc/eventhandler/webhookhandler.rst:24]: 604e4e6

File: doc/eventhandler/webhookhandler.rst [hunk 7]
```diff
@@ -21,4 +21,7 @@ what ever your other application understands as data.
...
+    "This webhook is triggered by John".
+    Possible please holder are `logged_in_user`, `realm`, `token_owner`, `user_realm` and `token_serial`
```
[the remark this answers is not in the document]
Comment (jona-samuel) [reply doc/eventhandler/webhookhandler.rst]: 604e4e6

File: privacyidea/lib/utils/__init__.py [hunk 8]
```diff
@@ -50,6 +50,7 @@
 import html
 
 from privacyidea.lib.error import ParameterError, ResourceNotFoundError, PolicyError
+from collections import defaultdict
```
Comment (plettich) [inline privacyidea/lib/utils/__init__.py:53]: This isn't needed anymore

File: privacyidea/lib/utils/__init__.py [hunk 9]
```diff
@@ -1480,3 +1481,41 @@ def parse_string_to_dict(s, split_char=":"):
...
+            "realm": realm,
+            "surname": surname,
+            "token_owner": givenname,
+            "user_realm": userrealm,
+            "token_serial": token_serial
+        }
+        new_text = text.format(**attributes)
+        return new_text
+    except(ValueError, KeyError) as err:
+        log.warning('privacyIDEA can`t replace your placeholder. Please check the text for mistakes')
```
Comment (plettich) [inline privacyidea/lib/utils/__init__.py:1520]: You can add the error message to the warning, makes it easier to figure out, which placeholder was wrong.

**Review by plettich — CHANGES_REQUESTED** [review]

File: privacyidea/lib/eventhandler/webhookevanthandler.py [hunk 10]
```diff
@@ -22,13 +22,16 @@
 api requests and reduce your traffic this way.
 
 """
+
```
[the remark this answers is not in the document]
Comment (jona-samuel) [reply privacyidea/lib/eventhandler/webhookevanthandler.py:25]: rename in?

[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/eventhandler/webhookevanthandler.py:25]: webhookev**a**nthandler

File: privacyidea/lib/utils/__init__.py (hunk 9 above)
Comment (cornelinux) [reply privacyidea/lib/utils/__init__.py:1520]: maybe also add a 

    log.debug(traceback...))

Comment (jona-samuel) [reply privacyidea/lib/utils/__init__.py]: 09f984a

File: tests/test_lib_utils.py [hunk 11]
```diff
@@ -24,7 +24,7 @@
                                    modhex_decode, checksum, urlsafe_b64encode_and_unicode,
                                    check_ip_in_policy, split_pin_pass, create_tag_dict,
                                    check_serial_valid, determine_logged_in_userparams,
-                                   to_list, parse_string_to_dict)
+                                   to_list, parse_string_to_dict, replace_function_event_handler)
```
[the remark this answers is not in the document]
Comment (jona-samuel) [reply tests/test_lib_utils.py]: 09f984a

File: privacyidea/lib/utils/__init__.py (hunk 8 above)
Comment (jona-samuel) [reply privacyidea/lib/utils/__init__.py]: 09f984a

File: privacyidea/lib/eventhandler/webhookevanthandler.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jona-samuel) [reply privacyidea/lib/eventhandler/webhookevanthandler.py:25]: 09f984a

**Review by plettich — APPROVED** [review]
